test/types/client_bulk_write.test-d.ts

```diff
@@ -0,0 +1,258 @@
+import { expectError, expectType } from 'tsd';
+
+import {
+  type ClientBulkWriteModel,
+  type ClientDeleteManyModel,
+  type ClientDeleteOneModel,
+  type ClientInsertOneModel,
+  type ClientReplaceOneModel,
+  type ClientUpdateManyModel,
+  type ClientUpdateOneModel,
+  type Document,
+  type Filter,
+  type MongoClient,
+  type OptionalId,
+  type UpdateFilter,
+  type UUID,
+  type WithoutId
+} from '../mongodb';
+
+declare const client: MongoClient;
+type Book = { title: string; released: Date };
+type Author = { name: string; published: number };
+type Store = { _id: UUID };
+
+// Baseline check that schema modifies the following fields for each type.
+expectType<ClientInsertOneModel<Book>['document']>(null as unknown as OptionalId<Book>);
+
+expectType<ClientReplaceOneModel<Book>['filter']>(null as unknown as Filter<Book>);
+expectType<ClientReplaceOneModel<Book>['replacement']>(null as unknown as WithoutId<Book>);
+
+expectType<ClientUpdateOneModel<Book>['filter']>(null as unknown as Filter<Book>);
+expectType<ClientUpdateOneModel<Book>['update']>(
+  null as unknown as UpdateFilter<Book> | Document[]
+);
+
+expectType<ClientUpdateManyModel<Book>['filter']>(null as unknown as Filter<Book>);
+expectType<ClientUpdateManyModel<Book>['update']>(
+  null as unknown as UpdateFilter<Book> | Document[]
+);
+
+expectType<ClientDeleteOneModel<Book>['filter']>(null as unknown as Filter<Book>);
+expectType<ClientDeleteManyModel<Book>['filter']>(null as unknown as Filter<Book>);
+
+client.bulkWrite([]); // empty should always work
+
+// No schemas - all correct use
+client.bulkWrite([
+  {
+    namespace: 'db.authors',
+    name: 'insertOne',
+    document: { name: 'bob', published: 2 }
+  },
+  {
+    namespace: 'db.authors',
+    name: 'replaceOne',
+    filter: { name: 'bob' },
+    replacement: { name: 'ann', published: 2 }
+  },
+  {
+    namespace: 'db.authors',
+    name: 'updateOne',
+    filter: { name: 'bob', published: 2 },
+    update: {}
+  },
+  {
+    namespace: 'db.authors',
+    name: 'updateMany',
+    filter: { name: 'bob', published: 2 },
+    update: {}
+  },
+  { namespace: 'db.authors', name: 'deleteOne', filter: {} },
+  { namespace: 'db.authors', name: 'deleteMany', filter: {} }
+]);
+
+// No schemas - incorrect use - random namespaces, no type checking
```

Comment:
```suggestion
// No schemas - random namespaces, no type checking
```

if we expect this to compile, its still correct usage.  it just isn't getting type safety